doc/examples/qsv_transcode.c

```diff
@@ -0,0 +1,431 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+/**
+ * @file
+ * Intel QSV-accelerated transcoding example.
+ *
+ * @example qsv_transcode.c
+ * This example shows how to do QSV-accelerated transcoding and how to
+ * dynamically change encoder's option.
+ * Usage: qsv_transcode input_stream codec output_stream initial option
+ *                      { frame_number new_option }
+ * e.g: - qsv_transcode input.mp4 h264_qsv output_h264.mp4 "g 60"
+ *      - qsv_transcode input.mp4 hevc_qsv output_hevc.mp4 "g 60 async_depth 1"
+ *                      100 "g 120"
+ *         (initialize codec with gop_size 60 and change it to 120 after 100
+ *          frames)
+ */
+
+#include <stdio.h>
+#include <errno.h>
+
+#include <libavutil/hwcontext.h>
+#include <libavcodec/avcodec.h>
+#include <libavformat/avformat.h>
+#include <libavutil/opt.h>
+
+static AVFormatContext *ifmt_ctx = NULL, *ofmt_ctx = NULL;
+static AVBufferRef *hw_device_ctx = NULL;
+static AVCodecContext *decoder_ctx = NULL, *encoder_ctx = NULL;
+static int video_stream = -1;
+static int initialized = 0;
+
+typedef struct DynamicSetting {
+    int frame_number;
+    char* optstr;
+} DynamicSetting;
+static DynamicSetting *dynamic_setting;
+static int setting_number;
+static int current_setting_number;
+
+static int str_to_dict(char* optstr, AVDictionary **opt)
+{
+    char *key, *value;
+    key = strtok(optstr, " ");
+    if (key == NULL)
+        return AVERROR(ENAVAIL);
+    value = strtok(NULL, " ");
+    if (value == NULL)
+        return AVERROR(ENAVAIL);
+    av_dict_set(opt, key, value, 0);
+    do {
+        key = strtok(NULL, " ");
+        if (key == NULL)
+            return 0;
+        value = strtok(NULL, " ");
+        if (value == NULL)
+            return AVERROR(ENAVAIL);
+        av_dict_set(opt, key, value, 0);
+    } while(key != NULL);
+    return 0;
+}
+
+static int dynamic_set_parameter(AVCodecContext *avctx)
+{
+    AVDictionary *opts = NULL;
+    int ret = 0;
+    static int frame_number = 0;
+    frame_number++;
+    if (current_setting_number < setting_number &&
+        frame_number == dynamic_setting[current_setting_number].frame_number) {
+        AVDictionaryEntry *e = NULL;
+        ret = str_to_dict(dynamic_setting[current_setting_number].optstr, &opts);
+        if (ret < 0) {
+            fprintf(stderr, "The dynamic parameter is wrong\n");
+            goto fail;
+        }
+        /* set comman option */
+        if ((ret = av_opt_set_dict(avctx, &opts)) < 0)
```

Comment:
What will happen if global context does not support this option? it will be just ignored? I would comment that in a code to help reader.

Reply:
I have added comment in the code. You can see the API document about av_opt_set_dict() as well.